### PR TITLE
VSX: Add 'Install Another Version...' Command

### PR DESCRIPTION
**Pull request by seantan22**
#### What it does

- Supports the ability to install any compatible version of an user-installed extension, provided that the extension is available in the Open VSX Registry.

https://user-images.githubusercontent.com/46289281/114419822-c3067980-9b81-11eb-9155-6556d10b7919.mp4

#### How to test

1. Open the extensions view and locate the `Installed` view of all user-installed extensions.
2. Pick an extension --> Click the `manage` gear icon --> Click `Install Another Version...` command. Note that the command is disabled for any extension not available in the Open VSX Registry.
3. Select a compatible version to install from the Quick Pick dropdown.
4. Observe that the previous version is uninstalled and replaced with the newly selected version.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: seantan22 <sean.a.tan@ericsson.com>

### CODE REVIEW

File: packages/plugin-ext/src/main/node/plugin-deployer-impl.ts [hunk 1]
```diff
@@ -132,14 +132,15 @@ export class PluginDeployerImpl implements PluginDeployer {
...
-        await this.deployMultipleEntries([pluginEntry], type);
+        await this.deployMultipleEntries([pluginEntry], type, options);
         this.logMeasurement('Deploy plugin entry', startDeployTime);
     }
 
-    protected async deployMultipleEntries(pluginEntries: ReadonlyArray<string>, type: PluginType = PluginType.System): Promise<void> {
-        const pluginsToDeploy = await this.resolvePlugins(pluginEntries, type);
+    protected async deployMultipleEntries(pluginEntries: ReadonlyArray<string>, type: PluginType = PluginType.System, options?: PluginDeployOptions): Promise<void> {
+        let pluginsToDeploy = [];
+        pluginsToDeploy = await this.resolvePlugins(pluginEntries, type, options);
```
Comment (colin-grant-work) [inline packages/plugin-ext/src/main/node/plugin-deployer-impl.ts:143]: Is this `let` declaration necessary, and if so, why? (Same question re: l. 179 below)

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 2]
```diff
@@ -24,12 +24,23 @@ import { ColorContribution } from '@theia/core/lib/browser/color-application-con
...
+import { MessageService, Mutable } from '@theia/core/lib/common';
 import { FileDialogService, OpenFileDialogProps } from '@theia/filesystem/lib/browser';
 import { LabelProvider } from '@theia/core/lib/browser';
 import { VscodeCommands } from '@theia/plugin-ext-vscode/lib/browser/plugin-vscode-commands-contribution';
-import { VSXExtensionsContextMenu, VSXExtension } from './vsx-extension';
+import { MenuContribution, MenuModelRegistry } from '@theia/core/lib/common/menu';
+import { VSXExtension, VSXExtensionsContextMenu } from './vsx-extension';
+import { QuickPickItem, QuickPickService } from '@theia/core/lib/common/quick-pick-service';
+import { VSXRegistryAPI } from '../common/vsx-registry-api';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:35]: Better just to add it to the `package.json`? Ask Paul - he's the expert on extraneous dependencies :-)

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 3]
```diff
@@ -96,6 +113,11 @@ export class VSXExtensionsContribution extends AbstractViewContribution<VSXExten
             execute: () => this.installFromVSIX()
         });
 
+        commands.registerCommand({ id: VSXExtensionsCommands.INSTALL_ANOTHER_VERSION.id }, {
+            execute: async (extension: VSXExtension) => this.installAnotherVersion(extension),
+            isEnabled: (extension: VSXExtension) => !extension.builtin && !!extension.downloadUrl
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:118]: Maybe a small comment that checking for `downloadUrl` guarantees it was installed from the / a VSX registry.

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 4]
```diff
@@ -24,12 +24,23 @@ import { ColorContribution } from '@theia/core/lib/browser/color-application-con
...
+import { VSXExtension, VSXExtensionsContextMenu } from './vsx-extension';
+import { QuickPickItem, QuickPickService } from '@theia/core/lib/common/quick-pick-service';
+import { VSXRegistryAPI } from '../common/vsx-registry-api';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import * as moment from 'moment';
+import { PluginServer } from '@theia/plugin-ext/lib/common';
 import { ClipboardService } from '@theia/core/lib/browser/clipboard-service';
+import { VSXExtensionRaw } from '../common/vsx-registry-types';
+
+interface QuickPickVersionItem {
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:41]: Same interface as `PluginDeployOptions` - maybe just reuse it?

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 5]
```diff
@@ -215,11 +242,66 @@ export class VSXExtensionsContribution extends AbstractViewContribution<VSXExten
...
     }
+
+    /**
+     * Updates an extension to a specific version.
+     *
+     * @param extension the extension to update.
+     * @param updateToVersion the version to update to.
+     * @param revertToVersion the version to revert to (in case of failure).
+     */
+    private async updateVersion(extension: VSXExtension, updateToVersion: string, revertToVersion: string): Promise<void> {
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:295]: Maybe some message service toasts here if the process fails at any step where the user might want to know? I'm not 100% sure about what messages may be displayed by the functions called here, but it seems like the user might want to know if they ended up with the old version again, for example.

File: packages/vsx-registry/src/browser/vsx-extensions-model.ts [hunk 6]
```diff
@@ -164,16 +164,21 @@ export class VSXExtensionsModel {
...
                     currInstalled.add(extension.id);
-                    refreshing.push(this.refresh(id));
+                    refreshing.push(this.refresh(id, version));
                 }
             }
             for (const id of this._installed) {
-                refreshing.push(this.refresh(id));
+                const extension = this.getExtension(id);
+                if (!extension) {
+                    return;
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/browser/vsx-extensions-model.ts:179]: Do we want to `return` here, or `continue` to the next extension?

File: packages/vsx-registry/src/node/vsx-extension-resolver.ts [hunk 7]
```diff
@@ -43,23 +44,38 @@ export class VSXExtensionResolver implements PluginDeployerResolver {
...
-            return;
-        }
-        if (extension.error) {
-            throw new Error(extension.error);
+
+        let extensionVersion: string | undefined;
+        let extension: VSXExtensionRaw | undefined;
+        if (options) {
+            extensionVersion = options.version;
+            console.log(`[${id}]: trying to resolve version ${extensionVersion}...`);
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/node/vsx-extension-resolver.ts:57]: Do we want to keep these logs (here and l. 66)?

File: packages/vsx-registry/src/node/vsx-extension-resolver.ts [hunk 8]
```diff
@@ -43,23 +44,38 @@ export class VSXExtensionResolver implements PluginDeployerResolver {
...
+            extension = await this.api.getExtension(id, { extensionVersion: extensionVersion });
+            if (!extension) {
```
Comment (colin-grant-work) [inline packages/vsx-registry/src/node/vsx-extension-resolver.ts:59]: The error handling is duplicated here. Could we have the `if (options)` block only cover down to `extension = await ...`  code, and then run the error handling?

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 9]
```diff
@@ -173,6 +190,43 @@ export class VSXExtensionsContribution extends AbstractViewContribution<VSXExten
...
+        let compatibleExtensions: VSXExtensionRaw[] = [];
+        if (latestCompatible) {
+            compatibleExtensions = extensions.slice(extensions.findIndex(ext => ext.version === latestCompatible.version));
+        }
+        const items: QuickPickItem[] = [];
+        compatibleExtensions.forEach(ext => {
+            let publishedDate = moment(ext.timestamp).fromNow();
+            if (currentVersion === ext.version) {
+                publishedDate += ' (Current)';
+            }
```
Comment (vince-fugnitto) [inline packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:213]: For some reason I do not see `current` added as a description to the current extension version:

https://user-images.githubusercontent.com/40359487/140516403-4efc9849-cce1-4fca-b950-c315833b1841.mp4

I also verified when it is not the first entry in the list.

**Review by vince-fugnitto — COMMENTED** [review]
The changes overall look good to me, only had a minor comment about the quick-input.

**Comment by colin-grant-work** [comment]
@vince-fugnitto, thanks for pointing out the `(Current)` issue. It got knocked out by some changes from Rob Moran's tooltip commit. You mentioned that the current version wasn't first - I don't think it should be, necessarily:

image

Earlier, you also expressed some misgivings about the mid-stream undeployment and redeployment. It seems as though VSCode pretty much always prompts you to reload if you want to change plugin versions. Should we just do the same, adding a button to reload if the plugin is awaiting deployment? In Electron that's fairly straightforward - I'm not quite sure what we should do in the Browser, since it would be expecting a full shutdown and restart of the back end.

**Comment by vince-fugnitto** [comment]
> You mentioned that the current version wasn't first - I don't think it should be, necessarily. 

That's right, not sure why I mentioned it should be the first in the list, I believe I just wanted to express it should be added to the currently deployed version :+1: 

> Earlier, you also expressed some misgivings about the mid-stream undeployment and redeployment. It seems as though VSCode pretty much always prompts you to reload if you want to change plugin versions. Should we just do the same, adding a button to reload if the plugin is awaiting deployment? In Electron that's fairly straightforward - I'm not quite sure what we should do in the Browser, since it would be expecting a full shutdown and restart of the back end.

I'm not sure either. Are there cases where a plugin will not be updated until the app is reloaded?

**Comment by colin-grant-work** [comment]
@tsmaeder, would you be willing to weigh in here, since y'all at Red Hat probably have the best understanding of the plugin system. Is it / should it be safe to uninstall one version of a plugin and start a different version in the same session, or would it be preferable to restart the whole application when swapping out versions of plugins?

**Comment by tsmaeder** [comment]
@colin-grant-work I have two observations:

1. I didn't see any code that adds a plugin to a running plugin host process, so we'll have to at least restart the plugin host, which is usually done by reloading the browser window.
2. Adding a plugin should not be a problem (we do it in "plugin dev mode"), but removing one may leave code running in existing plugin hosts that relies on resources that might not be there anymore. Think of an icon that has been removed in a newer version of the plugin. So I think _replacing_ a plugin requires the back end to be restarted, because we don't have a separate set of deployed plugins per plugin host.

**Comment by colin-grant-work** [comment]
Planning to pick this back up, but ran into [an issue](https://github.com/eclipse/openvsx/issues/431) with the queries to Open VSX. I'll try to work around it - I'm not sure we need the query that's failing.

**Comment by colin-grant-work** [comment]
@JonasHelming, this one has been waiting on a solution for issues with plugin uninstallation: #11084. Once that's merged, I can finish this up straightforwardly.

**Pull request closed without merging**